src/bootstrap/replication/peer_discovery.rs

```diff
@@ -0,0 +1,161 @@
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+use std::time::{Duration, Instant};
+
+use futures::StreamExt;
+use libp2p::identity::ed25519::Keypair;
+use libp2p::swarm::SwarmEvent;
+use libp2p::PeerId;
+use prost::Message as _;
+use serde_json;
+use tracing::{info, warn};
+
+use crate::bootstrap::replication::error::BootstrapError;
+use crate::bootstrap::replication::rpc_client::RpcClientsManager;
+use crate::cfg::DEFAULT_RPC_PORT;
+use crate::network::gossip::{Config as GossipConfig, SnapchainGossip};
+use crate::proto::{self, ContactInfoBody};
+use crate::utils::statsd_wrapper::StatsdClientWrapper;
+
+const INCOMPATIBLE_PEER_COOLDOWN: Duration = Duration::from_secs(30 * 60); // 30 minutes
+
+// PeerId -> time we marked it incompatible
+type CooldownMap = Arc<Mutex<HashMap<PeerId, Instant>>>;
+
+pub struct PeerDiscoverer {
+    gossip: SnapchainGossip,
+    rpc_manager: Arc<RpcClientsManager>,
+    target_height: u64,
+    incompatible_peers: CooldownMap,
+    shutdown_rx: tokio::sync::oneshot::Receiver<()>,
+}
+
+impl PeerDiscoverer {
+    pub async fn new(
+        gossip_config: &GossipConfig,
+        rpc_manager: Arc<RpcClientsManager>,
+        target_height: u64,
+        fc_network: crate::proto::FarcasterNetwork,
+        statsd_client: StatsdClientWrapper,
+        shutdown_rx: tokio::sync::oneshot::Receiver<()>,
+    ) -> Result<Self, BootstrapError> {
+        // Ephemeral keypair for discovery session
+        let keypair = Keypair::generate();
+        let gossip = SnapchainGossip::create(
```

Comment:
If we start up a regular gossip system
(1) Will we be advertising our own contact infos? What will happen if peers try to connect to us before initial sync is complete? 
(2) Will we be subscribed to decided values? This is unnecessary but isn't an actual problem, I just want to make sure I understand the behavior of the system.

Reply:
I didn't want to start a full gossip system because we don't have a lot of data that will be asked to serve - Like if we get sync requests for blocks or shard chunks, we don't have them, so we can't serve them/return errors, which might cause peers to blacklist us, and we havent even started yet. 

We dont handle decided values until after replication is done and we start the full gossip system.